patches/MDVA-26795__fix_performance_issue_in_cache_locking_mechanism__2.3.4.patch

```diff
@@ -0,0 +1,835 @@
++ */
++class Redis extends \Cm_Cache_Backend_Redis
++{
++    /**
++     * Local state of preloaded keys.
++     *
++     * @var array
++     */
++    private $preloadedData = [];
++
++    /**
++     * Array of keys to be preloaded.
++     *
++     * @var array
++     */
++    private $preloadKeys = [];
++
++    /**
++     * @param array $options
++     */
++    public function __construct($options = [])
++    {
++        $this->preloadKeys = $options['preload_keys'] ?? [];
++        parent::__construct($options);
++    }
++
++    /**
++     * Load value with given id from cache
++     *
++     * @param  string  $id                     Cache id
++     * @param  boolean $doNotTestCacheValidity If set to true, the cache validity won't be tested
++     * @return bool|string
++     */
++    public function load($id, $doNotTestCacheValidity = false)
++    {
++        if (!empty($this->preloadKeys) && empty($this->preloadedData)) {
++            $redis =  $this->_slave ?? $this->_redis;
++            $redis = $redis->pipeline();
++
++            foreach ($this->preloadKeys as $key) {
++                $redis->hGet(self::PREFIX_KEY . $key, self::FIELD_DATA);
++            }
++
++            $this->preloadedData = array_filter(array_combine($this->preloadKeys, $redis->exec()));
++        }
++
++        if (isset($this->preloadedData[$id])) {
++            return $this->_decodeData($this->preloadedData[$id]);
++        }
++
++        return parent::load($id, $doNotTestCacheValidity);
++    }
++
++    /**
++     * Cover errors on save operations, which may occurs when Redis cannot evict keys, which is expected in some cases.
++     *
++     * @param string $data
++     * @param string $id
++     * @param array $tags
++     * @param bool $specificLifetime
++     * @return bool
++     */
++    public function save($data, $id, $tags = [], $specificLifetime = false)
++    {
++        try {
++            parent::save($data, $id, $tags, $specificLifetime);
++        } catch (\Throwable $exception) {
++            return false;
++        }
++
++        return true;
++    }
++}
+diff -Nuar a/vendor/magento/framework/Cache/Backend/RemoteSynchronizedCache.php b/vendor/magento/framework/Cache/Backend/RemoteSynchronizedCache.php
+--- a/vendor/magento/framework/Cache/Backend/RemoteSynchronizedCache.php
++++ b/vendor/magento/framework/Cache/Backend/RemoteSynchronizedCache.php
+@@ -9,8 +9,10 @@ namespace Magento\Framework\Cache\Backend;
+ /**
+  * Remote synchronized cache
+  *
+- * This class created for correct work local caches with multiple web nodes,
+- * that will be check cache status from remote cache
++ * This class created for correct work witch local caches and multiple web nodes,
++ * in order to be sure that we always have up to date local version of cache.
++ * This class will be check cache version from remote cache and in case it newer
++ * than local one, it will update local one from remote cache a.k.a two level cache.
```

Comment:
Please don't use acronyms in comments - some non-native English speakers might not know what it means